File: neetcode/blind75/intervals/meeting_schedule_ii.py
# https://neetcode.io/problems/meeting-schedule-ii

"""
Definition of Interval:
class Interval(object):
    def __init__(self, start, end):
        self.start = start
        self.end = end
"""

class Solution:
    def minMeetingRooms(self, intervals: List[Interval]) -> int:
        if not intervals:
            return 0

        time = []

        for i in intervals:
            start, end = i.start, i.end
            time.append((start, 1))
            time.append((end, -1))

        # Python sorts tuples and lists like these lexicographically; compare the first element, and only if that doesn't differ, compare the second element, etc.
        time.sort(key=lambda t: (t[0], t[1]))
        # by sorting like this, if there is a conflict, there will be >= 2 t[1]s next to eachother. While if no conflict in that interval, a t[1] = 1 will be closed by a t[1] = -1 right after.

        print(time)

        count = 0
        max_days = 0

        for t in time:
            count += t[1]
            max_days = max(max_days, count)

        return max_days

File: neetcode/blind75_JS/Intervals/meeting-schedule-ii.js
// https://neetcode.io/problems/meeting-schedule-ii

/**
 * Definition of Interval:
 * class Interval {
 *   constructor(start, end) {
 *     this.start = start;
 *     this.end = end;
 *   }
 * }
 */

class Solution {
    /**
     * @param {Interval[]} intervals
     * @returns {number}
     */
    minMeetingRooms(intervals) {
        if (intervals.length < 2) {
            return 1
        }

        const times = []
        for (let i of intervals) {
            times.push([i.start, 1])
            times.push([i.end, -1])
        }

        times.sort((a,b) => {return (a === b ? a[1] - b[1] : a[0] - b[0])})
        console.log(times)

        let res = 0, count = 0;
        for (const t of times) {
            count += t[1];
            res = Math.max(res, count);
        }
        return res;
    }
}
